subsys/ipc/ipc_service/backends/ipc_icmsg_me_follower.c

```diff
@@ -0,0 +1,375 @@
+	}
+
+exit:
+	k_mutex_unlock(&data->epts_mutex);
+	return r;
+}
+
+static int send(const struct device *instance, void *token,
+		const void *msg, size_t len)
+{
+	const struct icmsg_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	ept_id_t *id = token;
+	int r;
+	int sent_bytes;
+
+	if (*id == INVALID_EPT_ID) {
+		return -ENOTCONN;
+	}
+
+	if (len >= SEND_BUF_SIZE - sizeof(ept_id_t)) {
+		return -EBADMSG;
+	}
+
+	k_mutex_lock(&dev_data->send_mutex, K_FOREVER);
+
+	/* TODO: Optimization: How to avoid this copying? */
+	/* Scatter list supported by icmsg? */
+	dev_data->send_buffer[0] = *id;
+	memcpy(dev_data->send_buffer + sizeof(ept_id_t), msg, len);
+
+	r = icmsg_send(conf, &dev_data->icmsg_data, dev_data->send_buffer,
+			len + sizeof(ept_id_t));
+
+	if (r > 0) {
+		sent_bytes = r - 1;
+	}
+
+	k_mutex_unlock(&dev_data->send_mutex);
+
+	if (r > 0) {
+		return sent_bytes;
+	} else {
+		return r;
+	}
+}
+
+const static struct ipc_service_backend backend_ops = {
+	.open_instance = open,
+	.register_endpoint = register_ept,
+	.send = send,
+};
+
+static int backend_init(const struct device *instance)
+{
+	const struct icmsg_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+
+	k_event_init(&dev_data->event);
+	k_mutex_init(&dev_data->epts_mutex);
+	k_mutex_init(&dev_data->send_mutex);
+
+	return icmsg_init(conf, &dev_data->icmsg_data);
+}
+
+#define BACKEND_CONFIG_POPULATE(i)						\
+	{									\
+		.tx_shm_size = DT_REG_SIZE(DT_INST_PHANDLE(i, tx_region)),	\
+		.tx_shm_addr = DT_REG_ADDR(DT_INST_PHANDLE(i, tx_region)),	\
+		.rx_shm_size = DT_REG_SIZE(DT_INST_PHANDLE(i, rx_region)),	\
+		.rx_shm_addr = DT_REG_ADDR(DT_INST_PHANDLE(i, rx_region)),	\
+		.mbox_tx = MBOX_DT_CHANNEL_GET(DT_DRV_INST(i), tx),		\
+		.mbox_rx = MBOX_DT_CHANNEL_GET(DT_DRV_INST(i), rx),		\
+	}
+
+#define DEFINE_BACKEND_DEVICE(i)						\
+	static const struct icmsg_config_t backend_config_##i =			\
+		BACKEND_CONFIG_POPULATE(i);					\
+	static struct backend_data_t backend_data_##i;				\
+										\
+	DEVICE_DT_INST_DEFINE(i,						\
+			 &backend_init,						\
+			 NULL,							\
+			 &backend_data_##i,					\
+			 &backend_config_##i,					\
+			 POST_KERNEL,						\
+			 CONFIG_IPC_SERVICE_REG_BACKEND_PRIORITY,		\
+			 &backend_ops);
+
+DT_INST_FOREACH_STATUS_OKAY(DEFINE_BACKEND_DEVICE)
+
+#if defined(CONFIG_IPC_SERVICE_BACKEND_ICMSG_ME_SHMEM_RESET)
+#define BACKEND_CONFIG_DEFINE(i) BACKEND_CONFIG_POPULATE(i),
+static int shared_memory_prepare(const struct device *arg)
+{
+	const struct icmsg_config_t *backend_config;
+	const struct icmsg_config_t backend_configs[] = {
+		DT_INST_FOREACH_STATUS_OKAY(BACKEND_CONFIG_DEFINE)
+	};
+
+	for (backend_config = backend_configs;
+	     backend_config < backend_configs + ARRAY_SIZE(backend_configs);
+	     backend_config++) {
+		icmsg_clear_tx_memory(backend_config);
+	}
+
+	return 0;
+}
+
+SYS_INIT(shared_memory_prepare, PRE_KERNEL_1, 1);
```

Comment:
I have not checked - will it happen before the netcore is booted?

Reply:
Yes, it will. It fixed some race conditions I encountered on device resets while testing.